atomix/cluster/src/test/java/io/atomix/raft/ControllableRaftContexts.java

```diff
@@ -646,6 +647,10 @@ static class DataLossChecker implements AppendListener {
 
     @Override
     public void onWrite(final IndexedRaftLogEntry indexed) {
+      final long index = indexed.index();
+      final var entryChecksum = indexed.getPersistedRaftRecord().checksum();
+      indexToChecksumMap.computeIfAbsent(index, (i) -> new ArrayList<>());
+      indexToChecksumMap.get(index).add(entryChecksum);
       delegate.onWrite(indexed);
     }
```

Comment:
This is the only change, which makes me a bit worried. Please take a look at whether we are fine with this, we need to find a way to remove the object from the listener.